cirq-core/cirq/devices/noise_utils.py

```diff
@@ -78,20 +79,21 @@ def __str__(self):
         return f'{self.gate_type}{self.qubits}'
 
     def __repr__(self) -> str:
-        fullname = f'{self.gate_type.__module__}.{self.gate_type.__qualname__}'
         qubits = ', '.join(map(repr, self.qubits))
-        return f'cirq.devices.noise_utils.OpIdentifier({fullname}, {qubits})'
+        return f'cirq.devices.noise_utils.OpIdentifier({proper_repr(self.gate_type)}, {qubits})'
 
     def _value_equality_values_(self) -> Any:
         return (self.gate_type, self.qubits)
 
     def _json_dict_(self) -> Dict[str, Any]:
-        gate_json = protocols.json_cirq_type(self._gate_type)
-        return {'gate_type': gate_json, 'qubits': self._qubits}
+        if hasattr(self.gate_type, '__name__'):
+            return {'gate_type': protocols.json_cirq_type(self.gate_type), 'qubits': self._qubits}
+        return {'gate_type': self._gate_type, 'qubits': self._qubits}
```

Comment:
Similar comments like the `repr`: it seems like this `if else` is handling an edge case because `protocols.json_cirq_type` is not handling certain inputs. What inputs can't `protocols.json_cirq_type` handle here?

Reply:
`self.gate_type` and `self._gate_type` seem to have the same value. If so, is this `if else` necessary?

Reply:
Inside the `if hasattr(self.gate_type, '__name__')`, the case where `gate_type` is a class (but not an instance), is handled, and the only thing returned is the name of said class. Outside the `if`, we give `self._gate_type` and this gate is converted into its JSON form. 

We could modify `protocols.json_cirq_type` like:

```python
def json_cirq_type(type_obj: Type) -> str:
    if hasattr(type_obj, '__name__'):
        namespace = json_namespace(type_obj)
        if namespace:
            return f'{namespace}.{type_obj.__name__}'
        return type_obj.__name__
    return type_obj
```

But then there are two possible return types.

> self.gate_type and self._gate_type seem to have the same value.

Yeah, not sure why I used one inside the if and the other outside, but the `if else` is still necessary to handle what I said above. Which one should be used?

Reply:
What do you think about

```python
def json_cirq_type(type_obj: Type) -> str:
    if hasattr(type_obj, '__name__'):
        namespace = json_namespace(type_obj)
        if namespace:
            return f'{namespace}.{type_obj.__name__}'
        return type_obj.__name__

    return type_obj.__class__.__name__
```

It shouldn't introduce breaking changes to `json_cirq_type`, only increasing it's functionality. We might also need to update the docstring of `json_cirq_type` from `it can be used to provide type-object serialization
    for classes that need it.` to `it can be used to provide type-object serialization for classes or instances that need it.`

Reply:
That would output:

```
{
  "cirq_type": "OpIdentifier",
  "gate_type": "SingleQubitCliffordGate",
  "qubits": [
    {
      "cirq_type": "LineQubit",
      "x": 0
    }
  ]
}
```

for 

```python
qubits = cirq.LineQubit.range(1)
cliff = cirq.SingleQubitCliffordGate.from_xz_map((cirq.X, False), (cirq.Z, False))
opid = cirq.OpIdentifier(cliff, qubits[0])
```

Which brings the problem of losing information back. I think this is because when the object itself is passed, some other part of `cirq.to_json` takes care of converting it to JSON, but if we do this we will only get the class name.

Reply:
Okay sounds good. This is fine to me them modulo one nit i've made above.